pkg/controller/controller_utils_test.go

```diff
@@ -209,162 +209,203 @@ func TestControllerExpectations(t *testing.T) {
-		KubeClient: clientset,
-		Recorder:   &record.FakeRecorder{},
+	tests := []test{
+		{name: "Replication controller with 2 replicas", numReplicas: 2},
+		{name: "Replication controller with 1 replica", numReplicas: 1},
+		{name: "Replication controller with no replicas", numReplicas: 0},
+		{name: "Replication controller with 5 replicas", numReplicas: 5},
 	}
 
-	controllerSpec := newReplicationController(1)
-	controllerRef := metav1.NewControllerRef(controllerSpec, v1.SchemeGroupVersion.WithKind("ReplicationController"))
+	prepareTests(tests)
+	shuffleTests(tests)
 
-	// Make sure createReplica sends a POST to the apiserver with a pod from the controllers pod template
-	err := podControl.CreatePods(context.TODO(), ns, controllerSpec.Spec.Template, controllerSpec, controllerRef)
-	assert.NoError(t, err, "unexpected error: %v", err)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
 
-	expectedPod := v1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Labels:       controllerSpec.Spec.Template.Labels,
-			GenerateName: fmt.Sprintf("%s-", controllerSpec.Name),
-		},
-		Spec: controllerSpec.Spec.Template.Spec,
+			assert.False(t, uidExp.SatisfiedExpectations(logger, test.rcKey),
+				"Controller %v satisfied expectations before deletion", test.rcKey)
+
+			for _, p := range test.rcPodNames {
+				uidExp.DeletionObserved(logger, test.rcKey, p)
+			}
+
+			assert.True(t, uidExp.SatisfiedExpectations(logger, test.rcKey),
+				"Controller %v didn't satisfy expectations after deletion", test.rcKey)
+
+			uidExp.DeleteExpectations(logger, test.rcKey)
+
+			assert.Nil(t, uidExp.GetUIDs(test.rcKey),
+				"Failed to delete uid expectations for %v", test.rcKey)
+		})
 	}
-	fakeHandler.ValidateRequest(t, "/api/v1/namespaces/default/pods", "POST", nil)
-	var actualPod = &v1.Pod{}
-	err = json.Unmarshal([]byte(fakeHandler.RequestBody), actualPod)
-	assert.NoError(t, err, "unexpected error: %v", err)
-	assert.True(t, apiequality.Semantic.DeepDerivative(&expectedPod, actualPod),
-		"Body: %s", fakeHandler.RequestBody)
 }
 
-func TestCreatePodsWithGenerateName(t *testing.T) {
+func TestCreatePods(t *testing.T) {
+
 	ns := metav1.NamespaceDefault
-	body := runtime.EncodeOrDie(clientscheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "empty_pod"}})
-	fakeHandler := utiltesting.FakeHandler{
-		StatusCode:   200,
-		ResponseBody: string(body),
-	}
-	testServer := httptest.NewServer(&fakeHandler)
-	defer testServer.Close()
-	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: testServer.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+	controllerSpec := newReplicationController(1)
+	controllerRef := metav1.NewControllerRef(controllerSpec, v1.SchemeGroupVersion.WithKind("ReplicationController"))
 
-	podControl := RealPodControl{
-		KubeClient: clientset,
-		Recorder:   &record.FakeRecorder{},
+	type test struct {
```

Comment:
I did it this way because of the [second test case](https://github.com/kaisoz/kubernetes/blob/02aca97542f83229b15b7b1e1c4aeb017e7e7172/pkg/controller/controller_utils_test.go#L362). As you can see, the `generateName` variable is used both in the `podCreationFunc` function and the `expectedPod` substruct.

The main problem is that, in Golang you can't reference a struct member from another struct member at struct definition time. So if I added the `generateName` variable to the test struct declaration, I'd get a compilation error when referencing it from both the `podCreationFunc` function and the `expectedPod` substruct.

I could define the `generateName` variable outside of the test array definition but that feels uncoherent because it's test input data and not logic.

Another solution would be to split the function again (which would make sense) and have one for testing the pod creation, and another to test the pod creation with name.

Does this make sense? WDYT?

Reply:
I’ve seen many examples of defining constants out of test cases and using those in struct defining.  I think that’s a good way to approach it.  A separate function seems heavy handed imo.